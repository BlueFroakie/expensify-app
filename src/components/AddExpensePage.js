import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {addExpense} from '../actions/expenses';

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        // props.dispatch(addExpense(expense));
        this.props.addExpense(expense); //We do this because this way the component is more testable, the addExpense function would be difficult to test.
        this.props.history.push('/');
    };

    render() {
        return (
                <div>
                    <h1>Add expense</h1>
                    <ExpenseForm 
                        onSubmit={this.onSubmit}
                    />   
                </div>
            );
        }
    };

const mapDispatchToProps = (dispatch) => ({
        addExpense: (expense) => dispatch(addExpense(expense))  
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);

/* For some reason the parenthesis after return inside of render has to be inline with the return keyword, otherwise the jsx is not recognized */