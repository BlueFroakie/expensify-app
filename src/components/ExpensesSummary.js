import React from 'react';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';
import {connect} from 'react-redux';
import {Link} from 'react-router-dom';

export const ExpensesSummary = ({expenseCount, expensesTotal}) => {
    const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
    const formattedExpensesTotal = numeral(expensesTotal / 100).format('$0,0.00')
    return (
        <div className="page-header">
            <div className="content-container">
                <h1 className="page-header__title">Viewing <span> {expenseCount} </span> {expenseWord} with a total of <span> {formattedExpensesTotal} </span></h1>
                <div className="page-header__actions">
                    <Link to="/create" className="button">Add expense</Link>                
                </div>
            </div>
        </div>
    );   
};

const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters)
    return {
        expenseCount: visibleExpenses.length,
        expensesTotal: selectExpensesTotal(visibleExpenses)
    };
};

export default connect(mapStateToProps)(ExpensesSummary);

// I used the props parameter on the stateless functional component but Andrew did it by destructuring the props object. 

