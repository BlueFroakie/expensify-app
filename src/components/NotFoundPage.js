import React from 'react';
import {Link} from 'react-router-dom';

const NotFoundPage = () => (
    <div>
    404!  <Link to={`/edit/${id}`}>Go home</Link>
    </div>
);

export default NotFoundPage;