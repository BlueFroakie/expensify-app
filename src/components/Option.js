import React from 'react';

const Option = (props) => {
    return (
        <div>
         {props.optionText}
         <a href="#" onClick={ (e) => {
             props.handleDeleteOption(props.optionText);
         }}>remove</a>
        </div>
        );
}

export default Option;