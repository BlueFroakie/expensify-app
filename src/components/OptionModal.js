import React from 'react';
import Modal from 'react-modal';

const OptionModal = (props) => (
<Modal 
    isOpen = {!!props.selectedOption}
    onRequestClose = {props.deselectOption}
    contentLabel = "Selected Option"
>
<h3>Selected option</h3>
{props.selectedOption && <p> {props.selectedOption} </p>}
<button onClick={props.deselectOption} > Ok </button>
</Modal>
);

export default OptionModal