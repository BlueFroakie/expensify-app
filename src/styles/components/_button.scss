.big-button {
    background: $cyan;
    border: none;
    border-bottom: .6rem solid darken($cyan, 10%);
    color: white;
    font-weight: bold;
    font-size: $l-size;
    margin-bottom: $m-size;
    padding: 2.4rem;
    width: 100%;
}

.big-button:disabled {
    opacity: .1;
}

.button {
    background: $cyan;
    border: none;
    border-bottom: .6rem solid darken($cyan, 10%);
    color: white;
    font-weight: 500;
    padding: $s-size;
}

.button--link {
    background: none;
    border: none;
    color: $off-white;
    padding: 0;
}

@media (min-width: $desktop-breakpoint) {
    .big-button {
        margin: 0 0 $xl-size 0;
    }
}