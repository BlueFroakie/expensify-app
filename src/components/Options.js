import React from 'react';
import Option from './Option';

const Options = (props) => (       
        <div>
        {props.options.length === 0 && <p>Add activities to get started</p>}
        {props.options.length === 1 && <p>Add at least one more activity to get started</p>}
        
        Your options here: <a onClick={props.handleRemoveAll} href="#">Remove all</a>
        <br/>
        ------------------
        {props.options.map((option) => <Option key = {option} optionText = {option} handleDeleteOption = {props.handleDeleteOption} />)}
        </div>
    );
    
export default Options;