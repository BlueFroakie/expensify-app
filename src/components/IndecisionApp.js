import React from 'react';
import AddOption from './AddOption';
import Options from './Options';
import Action from './Action';
import Header from './Header';

export default class IndecisionApp extends React.Component {
        state = {
            options : []
        }

        handlePick = () => {
            const randomNum = Math.floor(Math.random() * this.state.options.length);
            const optionChosen = this.state.options[randomNum];
            alert(optionChosen);
        }
    
        handleRemoveAll = () => {
            this.setState(() => ({ options: [] }));
        }
    
        handleDeleteOption = (optionToRemove) => {
            this.setState( (prevState) => ({ options: prevState.options.filter( option => optionToRemove !== option )}));
        }
    
        handleAddOption = (option) => {
            if (!option) {
                return 'Enter valid value to add item';
              } else if (this.state.options.indexOf(option) > -1) {
                return 'This option already exists';
              }
            this.setState((prevState) => ({options: prevState.options.concat(option)}));
        }

    componentDidMount() {
        try {
            const json = localStorage.getItem('options');
            const options = JSON.parse(json);

            if(options) {
                this.setState(() => ({options})); //options: options is the same as options
            }        
        } catch (e) {
            //do nothing
        }

    }

    componentDidUpdate(prevProps, prevState) {
        if(prevState.options.length !== this.state.options.length) {
            const json = JSON.stringify(this.state.options);
            localStorage.setItem('options', json);
        }
    }

    componentWillUnmount() {
        console.log(`Component will unmount`);
    }

    render() {
        const subtitle = `Let technology make your decisions`;

        return (
            <div>
            <Header subtitle = {subtitle} />
            <Action hasEnoughOptions = {this.state.options.length >= 2} handlePick = {this.handlePick} />
            <Options options = {this.state.options} handleRemoveAll = {this.handleRemoveAll} handleDeleteOption = {this.handleDeleteOption}/>
            <AddOption handleAddOption = {this.handleAddOption} />
            </div>
        );
    }
}
