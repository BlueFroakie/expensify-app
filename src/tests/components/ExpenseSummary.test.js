import React from 'react';
import {shallow} from 'enzyme';
import {ExpensesSummary} from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

test('Should render component with no expenses', () => {
    const wrapper = shallow(<ExpensesSummary expenseCount = {0} expensesTotal = {0}/>);
    expect(wrapper).toMatchSnapshot();
});

test('Should render component with one expense', () => {
    const wrapper = shallow(<ExpensesSummary expenseCount = {1} expensesTotal = {100}/>);
    expect(wrapper).toMatchSnapshot();
});

test('Should render component with two or more expenses', () => {
    const wrapper = shallow(<ExpensesSummary expenseCount = {2} expensesTotal = {10000}/>);
    expect(wrapper).toMatchSnapshot();
});