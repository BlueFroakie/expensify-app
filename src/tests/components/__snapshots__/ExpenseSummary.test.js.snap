// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render component with no expenses 1`] = `
<div>
  <h1>
    Viewing 
    0
     
    expenses
     with a total of 
    $0.00
  </h1>
</div>
`;

exports[`Should render component with one expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     with a total of 
    $1.00
  </h1>
</div>
`;

exports[`Should render component with two or more expenses 1`] = `
<div>
  <h1>
    Viewing 
    2
     
    expenses
     with a total of 
    $100.00
  </h1>
</div>
`;
